File: Java/LetterHistogram.java
package Java;

import java.util.Map;
import java.util.TreeMap;

public class LetterHistogram {
  private static void generateHistogram(String s) {
    Map<Character, Integer> map = new TreeMap<>();
    char[] array = s.replaceAll("[^a-zA-Z]", "").toLowerCase().toCharArray();

    for (char letter : array) {
      if (!map.containsKey(letter)) {
        map.put(letter, 1);
      } else {
        map.put(letter, map.get(letter) + 1);
      }
    }

    for (char key : map.keySet()) {
      System.out.printf("%c: %s\n", key, "=".repeat(map.get(key)));
    }
  }

  public static void main(String[] args) {
    generateHistogram("O rato roeu a roupa do rei de Roma");
  }
}

File: TS/letterHistogram.ts
function letterHistogram(text: string): void {
  let map = new Map<string, number>();

  for (let char of text) {
    if (!map.has(char)) {
      map.set(char, 1);
    } else {
      map.set(char, map.get(char)! + 1);
    }
  }

  for (let key of map.keys()) {
    console.log(`${key}: ${"=".repeat(map.get(key)!)}`);    
  }
}

function caseInsensitiveLetterHistogram(text: string): void {
  let map = new Map<string, number>();
  const cleanString = text.replace(/[^a-zA-Z]/g, "").toLowerCase();

  for (let char of cleanString) {
    if (!map.has(char)) map.set(char, 1);
    else map.set(char, map.get(char)! + 1);
  }

  for (let key of map.keys()) {
    console.log(key + ": " + "=".repeat(map.get(key)!));
    
  }
}

console.log(letterHistogram("O rato roeu a roupa do rei de Roma"));
console.log(caseInsensitiveLetterHistogram("O rato roeu a roupa do rei de Roma"));
